### ISSUE TITLE
no prefix, for loaded iri (truncated data)

### ISSUE THREAD

**Issue by JervenBolleman** [openlink/virtuoso-opensource#982]
In UniProt we sometimes get this error in the return from queries, and this specific case came in from our help desk. By a user which has loaded our data independently.
I am removing all sparql layers to show the issue. 
Currently it is showable live with [describe <http://purl.uniprot.org/EMBLWGS/MBP5512258>](https://sparql.uniprot.org/sparql/?query=describe+%3Chttp%3A%2F%2Fpurl.uniprot.org%2FEMBLWGS%2FMBP5512258%3E&format=html&query=describe+%3Chttp%3A%2F%2Fpurl.uniprot.org%2FEMBLWGS%2FMBP5512258%3E)

Bringing this down to the sql level we take the `no prefix` iri that was the `encodeBy` object.

```
SQL> select O from RDF_QUAD where S=iri_to_id('http://purl.uniprot.org/EMBLWGS/MBP5512258', 0) and P=iri_to_id('http://purl.uniprot.org/core/encodedBy', 0); 
O
LONG VARCHAR NOT NULL
_______________________________________________________________________________

#i18428523848

1 Rows. -- 2 msec.

```
Return that result back to an IRI string.

```
SQL> select id_to_iri(O) from RDF_QUAD where S=iri_to_id('http://purl.uniprot.org/EMBLWGS/MBP5512258', 0) and P=iri_to_id('http://purl.uniprot.org/core/encodedBy', 0); 
id_to_iri
LONG VARCHAR
_______________________________________________________________________________

no prefixgene-J6X90_02640

1 Rows. -- 2 msec.
SQL>
```
We lose the prefix part. 

Trying to figure out where this comes from I try

```
SQL> select id_to_iri(iri_id_from_num(18428523848));

*** Error 42000: [Virtuoso Driver][Virtuoso Server]SR673: Invalid IRI_ID #i18428523848
at line 44 of Top-Level:
select id_to_iri(iri_id_from_num(18428523848))
```

**Comment by TallTed** [comment]
@imitko  @pkleef @openlink -- Please take a look at this...

**Comment by imitko** [comment]
> *** Error 42000: [Virtuoso Driver][Virtuoso Server]SR673: Invalid IRI_ID #i18428523848
> at line 44 of Top-Level:
> select id_to_iri(iri_id_from_num(18428523848))

This error means no such IRI defined, i.e., the number does not have a corresponding record in `RDF_IRI` and/or `REF_PREFIX` table.

Therefore this error not a bug, it indicates a problem with `O` reported at the beginning about missing record for the IRI `ID` `18428523848`.

The solution is to find the corresponding data for that row in quad table `(G,S,P,O)`; delete that row from `RDF_QUAD`; and re-insert with SPARUL insert statement. If there are steps to reproduce missing IRI or prefix record, then open another issue.

**Comment by hideki-db** [comment]
The same server responds to the query —
`SELECT * { <http://purl.uniprot.org/EMBLWGS/MBP5512258> <http://purl.uniprot.org/core/encodedBy> ?o . }`
— with a wrong answer, `:no prefixgene-J6X90_02640`, but it responds to the query —
`SELECT * { <http://purl.uniprot.org/EMBLWGS/MBP5512258> ?p <http://purl.uniprot.org/EMBLWGS/MBP5512258#gene-J6X90_02640> . }`
— with the correct answer, `http://purl.uniprot.org/core/encodedBy`.

I think this means the server holds the correct triple.

We found the problems occur at IRI's with large `RP_ID` (`ID` in `DB.DBA.RDF_PREFIX` table), larger than `2^31-1`.

**Comment by JervenBolleman** [comment]
@imitko  For me the issue then should be renamed to `rdf_loader_run(log_enable=>2)` can leave `IRI_ID`s inserted in inconsistent state.

90 billion triples is a significant number to fix after the fact.

**Comment by imitko** [comment]
@JervenBolleman what is the average number of triples in the source RDF dumps? If not so many, then one always can specify `log_enable=>0` to disable row auto commit mode.

**Comment by natan-cox-cognizone** [comment]
Thank you, @imitko. I have a similar problem. 
What does "not so many" mean? Is 20 million still acceptable?

**Comment by pkleef** [comment]
A fix (b5e1ddbbef20a74f712d2d07604899e003a23881) has been committed to the `develop/7` branch and will be part of the upcoming new stable release.

This patch fixes two problems in Virtuoso:

1. When Virtuoso was upgraded to use 64-bit `IRI_ID`s around v6.0.x, we forgot to upgrade the `RDF_IRI` table to use a 64-bit prefix ID. This meant that Virtuoso could only store around 2 billion distinct prefixes before returning an error.
2. The algorithm to generate distinct prefixes resulted in too many prefixes being created.

While this is not a problem for storing small or even medium sized data sets, this becomes a problem when you want to host large databases the size of [Uniprot](https://www.uniprot.org/) which now contains over 90 billion triples.

To compare, here are some of the databases that OpenLink hosts, all of which are unaffected by this issue:

| Endpoint                                           | triples                 | distinct prefixes |
| -------------------------                | ------------:      | ---------------: |
| https://uriburner.com/sparql          | `138,881,702`        | `9,197,016`            |
| https://dbpedia.org/sparql              | `1,104,129,087`     | `27,528,113`         |
| https://wikidata.demo.openlinksw.com/sparql | `12,216,143,296` | `990,992` |
| https://lod.openlinksw.com/sparql | `35,875,699,899` |  `175,697,066`      |


The [`README.UPGRADE.md`](https://github.com/openlink/virtuoso-opensource/blob/develop/7/README.UPGRADE.md) contains more information on how to upgrade an existing database to the new 64-bit prefix IDs.
